### PR TITLE
Don't check main frame requests through shields network delegate helper

### PR DESCRIPTION
**Pull request by antonok-edm**
Resolves https://github.com/brave/brave-browser/issues/15047

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Visit https://shivankaul.com and click on the LinkedIn icon. The navigation to LinkedIn should not be blocked, both when `#brave-domain-block` is enabled and disabled under brave://flags.

Also, ensure that when `#brave-domain-block` is enabled under brave://flags, visiting https://1-1ads.com brings up a "Suspicious site ahead" interstitial with a "Proceed" button.

### CODE REVIEW

**Review by fmarier — APPROVED** [review]

File: browser/net/brave_ad_block_tp_network_delegate_helper.cc [hunk 1]
```diff
@@ -243,6 +243,13 @@ int OnBeforeURLRequest_AdBlockTPPreWork(const ResponseCallback& next_callback,
     return net::OK;
   }
 
+  // Requests for main frames are handled by DomainBlockNavigationThrottle,
+  // which can display a custom interstitial with an option to proceed if a
+  // block is made. We don't need to check these twice.
+  if (ctx->resource_type == blink::mojom::ResourceType::kMainFrame) {
```
Comment (iefremov) [inline browser/net/brave_ad_block_tp_network_delegate_helper.cc:249]: can we add a test?

Comment (antonok-edm) [reply browser/net/brave_ad_block_tp_network_delegate_helper.cc:249]: Added in https://github.com/brave/brave-core/pull/8402/commits/7d51927a0dad2211f2a07ff26b36afd0ce0ef65e

File: browser/brave_shields/domain_block_page_browsertest.cc [hunk 2]
```diff
@@ -299,6 +300,41 @@ IN_PROC_BROWSER_TEST_F(DomainBlockTest, NoFetch) {
...
+  NavigateTo(url);
+  ASSERT_FALSE(IsShowingInterstitial());
+
+  // Navigate to a page on the third-party b.com. There should be no
+  // interstitial shown.
+  std::string clickLink =
+      "domAutomationController.send(clickLink('" + cross_url.spec() + "'));";
+  bool success = false;
+  EXPECT_TRUE(
+      ExecuteScriptAndExtractBool(web_contents(), clickLink.c_str(), &success));
```
Comment (iefremov) [inline browser/brave_shields/domain_block_page_browsertest.cc:322]: maybe rewrite this with EvalJs?

File: browser/brave_shields/domain_block_page_browsertest.cc [hunk 3]
```diff
@@ -299,6 +300,41 @@ IN_PROC_BROWSER_TEST_F(DomainBlockTest, NoFetch) {
...
+  // No interstitial should be shown, since top-level requests are never
+  // third-party.
+  ASSERT_FALSE(IsShowingInterstitial());
+
+  // The default "blocked by an extension" interstitial also should not be
+  // shown. This would appear if the request was blocked by the network delegate
+  // helper.
+  const std::string location =
+      EvalJs(web_contents(), "window.location.href").ExtractString();
+  ASSERT_STRNE("chrome-error://chromewebdata/", location.c_str());
```
Comment (iefremov) [inline browser/brave_shields/domain_block_page_browsertest.cc:335]: no need for c_str

**Review by iefremov — APPROVED** [review]

File: browser/brave_shields/domain_block_page_browsertest.cc (hunk 3 above)
Comment (antonok-edm) [reply browser/brave_shields/domain_block_page_browsertest.cc:335]: I think we do need it, when I remove it I get this:

```
../../third_party/googletest/src/googletest/include/gtest/gtest.h:1669:28: note: candidate function not viable: no known conversion from 'const std::string' (aka 'const basic_string<char>') to 'const char *' for 4th argument
GTEST_API_ AssertionResult CmpHelperSTRNE(const char* s1_expression,
                           ^
../../third_party/googletest/src/googletest/include/gtest/gtest.h:1694:28: note: candidate function not viable: no known conversion from 'const char [30]' to 'const wchar_t *' for 3rd argument
GTEST_API_ AssertionResult CmpHelperSTRNE(const char* s1_expression,
```

File: browser/brave_shields/domain_block_page_browsertest.cc (hunk 2 above)
Comment (antonok-edm) [reply browser/brave_shields/domain_block_page_browsertest.cc]: good call, updated

**Comment by kjozwiak** [comment]
Verification **PASSED** on `Win 10 x64` using the following build:

```
Brave | 1.24.50 Chromium: 90.0.4430.51 (Official Build) nightly (64-bit)
-- | --
Revision | 32e5fa33a31641bded70a90e60121060691e7125-refs/branch-heads/4430@{#927}
OS | Windows 10 OS Version 2009 (Build 19042.868)
```

Test Cases from https://github.com/brave/brave-core/pull/8402#issue-606075390:

* visited https://shivankaul.com and ensured it loaded without any issues with `#brave-domain-block` set as `Default` (enabled)
* visited https://shivankaul.com and ensured it loaded without any issues with `#brave-domain-block` set as `Enabled` (enabled)
* visited https://shivankaul.com and ensured it loaded without any issues with `#brave-domain-block` set as `Disabled`

**`Enabled Example`** | **`Enabled Example`**
-------------------------|---------------------------
enabled1 | enabled2

**`Disabled Example`** | **`Disabled Example`**
-------------------------|---------------------------
disabled1 | disabled2

Visited https://1-1ads.com and ensured that Brave brings up a "Suspicious site ahead" interstitial with a "Proceed" button as per the following:

domainBlocked

Test Cases from https://github.com/brave/brave-browser/issues/15047#issue-847003613:

Ensured that selecting the `Unsubscribe` button and clicking on `Submit` works as expected:

DowUnsuscribe

**Pull request merged**